dagql/server.go

```diff
@@ -0,0 +1,890 @@
+package dagql
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"reflect"
+	"sort"
+	"sync"
+
+	"github.com/99designs/gqlgen/graphql"
+	"github.com/dagger/dagger/dagql/idproto"
+	"github.com/iancoleman/strcase"
+	"github.com/opencontainers/go-digest"
+	"github.com/sourcegraph/conc/pool"
+	"github.com/vektah/gqlparser/v2/ast"
+	"github.com/vektah/gqlparser/v2/gqlerror"
+	"github.com/vektah/gqlparser/v2/parser"
+	"golang.org/x/sync/errgroup"
+)
+
+type AroundFunc func(
+	context.Context,
+	Object,
+	*idproto.ID,
+	func(context.Context) (Typed, error),
+) func(context.Context) (Typed, error)
+
+// Server represents a GraphQL server whose schema is dynamically modified at
+// runtime.
+type Server struct {
+	root        Object
+	telemetry   AroundFunc
+	objects     map[string]ObjectType
+	scalars     map[string]ScalarType
+	typeDefs    map[string]TypeDef
+	directives  map[string]DirectiveSpec
+	installLock *sync.Mutex
+
+	// Cache is the inner cache used by the server. It can be replicated to
+	// another *Server to inherit and share caches.
+	//
+	// TODO: copy-on-write
+	Cache Cache
+}
+
+// Cache stores results of pure selections against Server.
+type Cache interface {
+	GetOrInitialize(
+		context.Context,
+		digest.Digest,
+		func(context.Context) (Typed, error),
+	) (Typed, error)
+}
+
+// TypeDef is a type whose sole practical purpose is to define a GraphQL type,
+// so it explicitly includes the Definitive interface.
+type TypeDef interface {
+	Type
+	Definitive
+}
+
+// NewServer returns a new Server with the given root object.
+func NewServer[T Typed](root T) *Server {
+	rootClass := NewClass[T](ClassOpts[T]{
+		// NB: there's nothing actually stopping this from being a thing, except it
```

Comment:
To clarify, this is saying that `Query` itself isn't id-able right?

Reply:
Yep!